crypto/provider_core.c

```diff
@@ -654,13 +673,21 @@ static void provider_activate_fallbacks(struct provider_store_st *store)
             OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(store->providers, i);
 
             /*
-             * Note that we don't care if the activation succeeds or not.
-             * If it doesn't succeed, then any attempt to use any of the
-             * fallback providers will fail anyway.
+             * Activated fallback providers get an extra refcount, to
+             * simulate a regular load.
+             * Note that we don't care if the activation succeeds or not,
+             * other than to maintain a correct refcount.  If the activation
+             * doesn't succeed, then any future attempt to use the fallback
+             * provider will fail anyway.
              */
             if (prov->flag_fallback) {
                 activated_fallback_count++;
-                provider_activate(prov);
+
+                if (ossl_provider_up_ref(prov)
+                    && !provider_activate(prov))
+                    ossl_provider_free(prov);
+                else
+                    prov->flag_activated_as_fallback = 1;
```

Comment:
So we set this to 1 if `ossl_provider_up_ref(prov)` returns false? That doesn't seem quite right

Reply:
Ah, no, that logic wasn't quite right.  Fixed, I hope it makes more sense now.